File: libs/windows/extra/autostart.cpp
#include "autostart.h"

#include "../core/debug.h"
#include "../base/application.h"
#include <windows.h>

namespace Windows {

bool setProgramToAutostart(bool value) {
  HKEY key;
  LONG error = RegCreateKeyExW (
    HKEY_CURRENT_USER,
    L"Software\\Microsoft\\Windows\\CurrentVersion\\Run",
    0, NULL, REG_OPTION_NON_VOLATILE,
    KEY_ALL_ACCESS, NULL, &key, 0);
  win_throw_on_fail(error);

  if (value) {
    error = RegSetValueExW(key, Application::getName().c_str(), 0, REG_SZ,
      (LPBYTE)(Application::getExecutablePath().c_str()), (DWORD)((Application::getExecutablePath().size() + 1) * sizeof(wchar_t)));
  } else {
    if (RegQueryValueExW(key, Application::getName().c_str(),0,0,0,0) != ERROR_FILE_NOT_FOUND)
      error = RegDeleteValueW(key, Application::getName().c_str());
  }

  RegCloseKey(key); // TODO: use RAII
  win_throw_on_fail(error);
  return true;
}

bool isProgramInAutostart() {
  HKEY key;
  LONG error;

  error = RegCreateKeyExW(
    HKEY_CURRENT_USER,
    L"Software\\Microsoft\\Windows\\CurrentVersion\\Run",
    0, NULL, REG_OPTION_NON_VOLATILE,
    KEY_ALL_ACCESS, NULL, &key, 0);
  win_throw_on_fail(error);


  DWORD buffersize = 0;

#if (WINVER < _WIN32_WINNT_VISTA)
  DWORD type;
  error = RegQueryValueEx(key, Application::getName().c_str(), NULL,  &type, NULL, &buffersize);
#else
  error = RegGetValue(key, NULL, Application::getName().c_str(), RRF_RT_REG_SZ, NULL, NULL, &buffersize);
#endif

  if (error != ERROR_SUCCESS) {
    if (error != ERROR_FILE_NOT_FOUND) {
      WIN_WARNING(L"cannot read autostart registry value: %d", error);
    }
    RegCloseKey(key); // TODO: use RAII
    return false;
  }
#if (WINVER < _WIN32_WINNT_VISTA)
  if (type != REG_SZ) {
    WIN_WARNING(L"%s", L"autostart registry value has wrong type.");
    RegCloseKey(key); // TODO: use RAII
    return false;
  }
#endif

  wchar_t buffer[buffersize+1];
#if (WINVER < _WIN32_WINNT_VISTA)
  error = RegQueryValueEx(key, Application::getName().c_str(), NULL,  &type, (BYTE*)buffer, &buffersize);
#else
  error = RegGetValue(key, NULL, Application::getName().c_str(), RRF_RT_REG_SZ, NULL, buffer, &buffersize);
#endif

  if (error != ERROR_SUCCESS) {
    if (error != ERROR_FILE_NOT_FOUND) {
      win_throw_on_fail(error);
    }
    RegCloseKey(key); // TODO: use RAII
    return false;
  } else {
    RegCloseKey(key); // TODO: use RAII
    return (Application::getExecutablePath().compare(buffer) == 0);
  }
}

} // namespace Windows

File: lightports/src/shell/autostart.rs
use crate::error::WinResult;
use crate::{result, FromWide, WString};
use std::ffi::{c_void, OsStr, OsString};
use std::{io, ptr};
use winapi::_core::mem::MaybeUninit;
use winapi::shared::minwindef::HKEY;
use winapi::shared::winerror::{ERROR_FILE_NOT_FOUND, ERROR_SUCCESS};
use winapi::um::winnt::{KEY_ALL_ACCESS, REG_OPTION_NON_VOLATILE, REG_SZ};
use winapi::um::winreg::{
    RegCloseKey, RegCreateKeyExW, RegDeleteValueW, RegGetValueW, RegSetValueExW, HKEY_CURRENT_USER,
    RRF_RT_REG_SZ,
};

struct KeyHandle(HKEY);

impl Drop for KeyHandle {
    fn drop(&mut self) {
        unsafe {
            RegCloseKey(self.0);
        }
    }
}

unsafe fn open_autostart_key() -> io::Result<KeyHandle> {
    let mut key: MaybeUninit<HKEY> = MaybeUninit::uninit();
    result(RegCreateKeyExW(
        HKEY_CURRENT_USER,
        WString::from_os_str("Software\\Microsoft\\Windows\\CurrentVersion\\Run").as_ptr(),
        0,
        ptr::null_mut(),
        REG_OPTION_NON_VOLATILE,
        KEY_ALL_ACCESS,
        ptr::null_mut(),
        key.as_mut_ptr(),
        ptr::null_mut(),
    ))?;
    Ok(KeyHandle(key.assume_init()))
}

pub fn set_to_autostart(app_id: &str, app_path: &OsStr) -> io::Result<()> {
    unsafe {
        let key = open_autostart_key()?;
        let value = WString::from_os_str(app_path);

        RegSetValueExW(
            key.0,
            WString::from_os_str(app_id).as_ptr(),
            0,
            REG_SZ,
            value.as_ptr() as *const _ as *const u8,
            value.bytes_len() as u32,
        )
        .into_void_result()
    }
}

pub fn remove_from_autostart(app_id: &str) -> io::Result<()> {
    unsafe {
        let key = open_autostart_key()?;

        let error = RegDeleteValueW(key.0, WString::from_os_str(app_id).as_ptr());
        if error != ERROR_FILE_NOT_FOUND as i32 {
            Err(io::Error::from_raw_os_error(error))
        } else {
            Ok(())
        }
    }
}

pub fn get_autostart_path(app_id: &str) -> io::Result<Option<OsString>> {
    unsafe {
        let key = open_autostart_key()?;

        let mut buffer: Vec<u16> = vec![];
        let mut buffersize: u32 = 0;
        let mut error = RegGetValueW(
            key.0,
            ptr::null_mut(),
            WString::from_os_str(app_id).as_ptr(),
            RRF_RT_REG_SZ,
            ptr::null_mut(),
            ptr::null_mut(),
            &mut buffersize,
        );
        if error == ERROR_SUCCESS as i32 {
            buffer.resize(buffersize as usize, 0);
            error = RegGetValueW(
                key.0,
                ptr::null_mut(),
                WString::from_os_str(app_id).as_ptr(),
                RRF_RT_REG_SZ,
                ptr::null_mut(),
                buffer.as_mut_ptr() as *mut c_void,
                &mut buffersize,
            )
        }

        if error != ERROR_SUCCESS as i32 {
            if error == ERROR_FILE_NOT_FOUND as i32 {
                Ok(None)
            } else {
                Err(io::Error::from_raw_os_error(error))
            }
        } else {
            Ok(Some(OsString::from_wide(&buffer)))
        }
    }
}
